### PR TITLE
[Refactor]: URLAction and SwitchAccount Router

### PR DESCRIPTION
**Pull request by marcomaddy1982**
## What's new in this PR?

In order to decouple and make the `AppRootRouter` lighter I created two new different routers: the `URLActionRouter` and the `SwitchAccountRouter`. The first one implements the `URLActionDelegate` and it takes care of presenting on the `rootViewController` the necessary alert depending on the `URLActionDelegate` call back. The second one implements the `SessionManagerSwitchingDelegate` and it takes care to present on the `rootViewController` the alert for switching the user account.

### CODE REVIEW

File: Wire-iOS/Sources/URLActionRouter.swift [hunk 1]
```diff
@@ -80,32 +84,46 @@ extension AppRootViewController: URLActionDelegate {
...
         alert.addAction(agreeAction)
         
         let cancelAction = UIAlertAction(title: "general.cancel".localized, style: .cancel)
         alert.addAction(cancelAction)
         
-        self.present(alert, animated: true, completion: nil)
+        rootViewController.present(alert, animated: true, completion: nil)
+    }
+    
+    private func performSwitchBakend(with configurationURL: URL) {
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/URLActionRouter.swift:109]: Suggestion `switchBackend(with:)`

**Review by johnxnguyen — COMMENTED** [review]
Is this code covered by unit tests?

**Comment by marcomaddy1982** [comment]
> Is this code covered by unit tests?

No, also before were no covered.

**Comment by johnxnguyen** [comment]
> > Is this code covered by unit tests?
> 
> No, also before were no covered.

I think we should add coverage not only for this change but for the rest of the refactoring because refactoring gives us the opportunity to make it testable. Is this something you're planning to do / have ideas about how to do it?

**Comment by marcomaddy1982** [comment]
> > > Is this code covered by unit tests?
> > 
> > 
> > No, also before were no covered.
> 
> I think we should add coverage not only for this change but for the rest of the refactoring because refactoring gives us the opportunity to make it testable. Is this something you're planning to do / have ideas about how to do it?

Yes, right.

File: Wire-iOS/Sources/URLActionRouter.swift [hunk 2]
```diff
@@ -16,47 +16,51 @@
...
-import Foundation
-import UIKit
 import WireSyncEngine
 
 extension Notification.Name {
     static let companyLoginDidFinish = Notification.Name("Wire.CompanyLoginDidFinish")
 }
 
-extension AppRootViewController: URLActionDelegate {
+final class URLActionRouter {
```
Comment (typfel) [inline Wire-iOS/Sources/URLActionRouter.swift:25]: This currently doesn't  follow the VIPER architecture. Shouldn't the `URLActionDelegate` be implemented in a `Interactor` and the alert in a `Presenter`

**Review by typfel — CHANGES_REQUESTED** [review]

File: Wire-iOS/Sources/URLActionRouter.swift [hunk 3]
```diff
@@ -16,47 +16,51 @@
...
     
-    // MARK - Public Implementation
+    // MARK: - Private Property
+    private let rootViewController: RootViewController
```
[the remark this answers is not in the document]
Comment (marcomaddy1982) [reply Wire-iOS/Sources/URLActionRouter.swift:28]: I thinks that alert should be always presented from the `RootViewController`, do you have in mind some issues with it?

File: Wire-iOS/Sources/URLActionRouter.swift [hunk 4]
```diff
@@ -16,47 +16,51 @@
...
         }
     }
     
-    func shouldPerformAction(_ action: URLAction, decisionHandler: @escaping (Bool) -> Void) {
+    public func shouldPerformAction(_ action: URLAction, decisionHandler: @escaping (Bool) -> Void) {
         switch action {
         case .connectBot:
             presentConnectBotAlert(with: decisionHandler)
         case .accessBackend(configurationURL: let configurationURL):
-            guard SecurityFlags.customBackend.isEnabled else {
```
[the remark this answers is not in the document]
Comment (marcomaddy1982) [reply Wire-iOS/Sources/URLActionRouter.swift:48]: we can move it back or rename `presentCustomBackendAlert` in `presentCustomBackendAlertIfNecesssary`.

File: Wire-iOS/Sources/URLActionRouter.swift (hunk 2 above)
Comment (marcomaddy1982) [reply Wire-iOS/Sources/URLActionRouter.swift:25]: The idea of this class was decouple a bit the `AppRootRouter`, we could also add the `Interactor` and the `Presenter` but IMHO in this case the code would be just more verbose.

File: Wire-iOS/Sources/URLActionRouter.swift [hunk 5]
```diff
@@ -80,32 +84,46 @@ extension AppRootViewController: URLActionDelegate {
...
     private func presentCustomBackendAlert(with configurationURL: URL) {
         let alert = UIAlertController(title: "url_action.switch_backend.title".localized,
                                       message: "url_action.switch_backend.message".localized(args: configurationURL.absoluteString),
                                       preferredStyle: .alert)
+        guard SecurityFlags.customBackend.isEnabled else {
+            return
+        }
+        
         let agreeAction = UIAlertAction(title: "general.ok".localized, style: .default) { _ in
-            self.isLoadingViewVisible = true
```
[the remark this answers is not in the document]
Comment (marcomaddy1982) [reply Wire-iOS/Sources/URLActionRouter.swift:91]: Nice catch, Thanks

File: Wire-iOS/Sources/URLActionRouter.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcomaddy1982) [reply Wire-iOS/Sources/URLActionRouter.swift:28]: moved back to the old position

File: Wire-iOS/Sources/URLActionRouter.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcomaddy1982) [reply Wire-iOS/Sources/URLActionRouter.swift:48]: moved back to the old position

File: Wire-iOS/Sources/URLActionRouter.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (marcomaddy1982) [reply Wire-iOS/Sources/URLActionRouter.swift:91]: Done.

File: Wire-iOS/Sources/URLActionRouter.swift (hunk 2 above)
Comment (typfel) [reply Wire-iOS/Sources/URLActionRouter.swift:25]: Ok, we can review it that later. This refactoring is already big enough :)

**Review by typfel — APPROVED** [review]

**Pull request merged**